net/connUDP.go

```diff
@@ -208,88 +205,173 @@ func (c *UDPConn) Close() error {
 	return c.connection.Close()
 }
 
-func (c *UDPConn) writeToAddr(multicastHopLimit int, iface net.Interface, srcAddr net.Addr, port string, raddr *net.UDPAddr, buffer []byte) error {
-	netType := "udp4"
+func (c *UDPConn) writeToAddr(iface *net.Interface, src *net.IP, multicastHopLimit int, raddr *net.UDPAddr, buffer []byte) error {
+	var pktSrc net.IP
+	var p packetConn
 	if IsIPv6(raddr.IP) {
-		netType = "udp6"
+		p = newPacketConnIPv6(ipv6.NewPacketConn(c.connection))
+		pktSrc = net.IPv6zero
+	} else {
+		p = newPacketConnIPv4(ipv4.NewPacketConn(c.connection))
+		pktSrc = net.IPv4zero
 	}
-	addrMask := srcAddr.String()
-	addr := strings.Split(addrMask, "/")[0]
-	if strings.Contains(addr, ":") && netType == "udp4" {
-		return nil
+	if src != nil {
+		pktSrc = *src
 	}
-	if !strings.Contains(addr, ":") && netType == "udp6" {
-		return nil
+
+	if c.closed.Load() {
+		return ErrConnectionIsClosed
 	}
-	var p packetConn
-	if netType == "udp4" {
-		p = newPacketConnIPv4(ipv4.NewPacketConn(c.connection))
+	if iface != nil {
+		if err := p.SetMulticastInterface(iface); err != nil {
+			return err
+		}
+	}
+	if err := p.SetMulticastHopLimit(multicastHopLimit); err != nil {
+		return err
+	}
+
+	var err error
+	if iface != nil || src != nil {
+		_, err = p.WriteTo(buffer, &ControlMessage{
+			Src:     pktSrc,
+			IfIndex: iface.Index,
+		}, raddr)
 	} else {
-		p = newPacketConnIPv6(ipv6.NewPacketConn(c.connection))
+		_, err = p.WriteTo(buffer, nil, raddr)
 	}
+	return err
+}
 
-	if err := p.SetMulticastInterface(&iface); err != nil {
-		return err
+func filterAddressesByNetwork(network string, ifaceAddrs []net.Addr) []net.Addr {
+	filtered := make([]net.Addr, 0, len(ifaceAddrs))
+	for _, srcAddr := range ifaceAddrs {
+		addrMask := srcAddr.String()
+		addr := strings.Split(addrMask, "/")[0]
+		if strings.Contains(addr, ":") && network == "udp4" {
+			continue
+		}
+		if !strings.Contains(addr, ":") && network == "udp6" {
+			continue
+		}
+		filtered = append(filtered, srcAddr)
 	}
-	if err := p.SetMulticastHopLimit(multicastHopLimit); err != nil {
+	return filtered
+
+}
+
+func convAddrsToIps(ifaceAddrs []net.Addr) []net.IP {
+	ips := make([]net.IP, 0, len(ifaceAddrs))
+	for _, addr := range ifaceAddrs {
+		addrMask := addr.String()
+		addr := strings.Split(addrMask, "/")[0]
+		ip := net.ParseIP(addr)
+		if ip != nil {
+			ips = append(ips, ip)
+		}
+	}
+	return ips
+}
+
+func (c *UDPConn) WriteMulticast(ctx context.Context, raddr *net.UDPAddr, buffer []byte, opts ...MulticastOption) error {
```

Comment:
Add docu to public API, doesn't have to be long.

Reply:
Fixed